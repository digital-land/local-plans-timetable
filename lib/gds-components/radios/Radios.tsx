/**
 * GDS component: https://design-system.service.gov.uk/components/radios/
 */
import cn from "classnames";

export type RadioOption = { label: string; value: string };

type RadiosProps = {
  label?: string;
  onChange: (value: string) => void;
  options: RadioOption[];
  id: string;
  inline?: boolean;
  error?: string;
  selectedOption?: string;
};

export const Radios = ({
  label,
  onChange,
  inline = false,
  options,
  error,
  id,
  selectedOption,
}: RadiosProps): JSX.Element => {
  return (
    <div
      className={cn("govuk-form-group", { "govuk-form-group--error": error })}
    >
      <fieldset className="govuk-fieldset" id={id}>
        {label && (
          <legend className="govuk-fieldset__legend govuk-fieldset__legend--l">
            <h1 className="govuk-fieldset__heading">{label}</h1>
          </legend>
        )}
        {error && (
          <p className="govuk-error-message">
            <span className="govuk-visually-hidden">Error:</span> {error}
          </p>
        )}
        <div
          className={cn("govuk-radios", { "govuk-radios--inline": inline })}
          data-module="govuk-radios"
        >
          {options.map(({ label, value }) => (
            <div className="govuk-radios__item" key={label}>
              <input
                className="govuk-radios__input"
                type="radio"
                value={value}
                id={value}
                name="radios"
                role="radio-input"
                checked={selectedOption === value}
                onChange={(e) => onChange(e.target.value)}
              />
              <label
                className="govuk-label govuk-radios__label"
                htmlFor={value}
              >
                {label}
              </label>
            </div>
          ))}
        </div>
      </fieldset>
    </div>
  );
};
