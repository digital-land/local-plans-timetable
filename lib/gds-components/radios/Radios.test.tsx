import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { RadioOption, Radios } from "./Radios";

const options: RadioOption[] = [
  { label: "Yes", value: "y" },
  { label: "No", value: "n" },
];
const onChange = jest.fn();

describe("Radios", () => {
  it("calls onChange", async () => {
    const { asFragment } = render(
      <Radios label="radio" onChange={onChange} options={options} />
    );

    await userEvent.click(screen.getByLabelText(options[0].label));

    expect(asFragment()).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledWith(options[0].value);
  });
  
  it("displays 1 radio input for each option", async () => {
    const onChange = jest.fn();
    const { asFragment } = render(
      <Radios label="radio" onChange={onChange} options={options} />
    );

    const inputs = screen.queryAllByRole("radio-input");

    expect(asFragment()).toMatchSnapshot();
    expect(inputs.length).toEqual(options.length);
  });
});
