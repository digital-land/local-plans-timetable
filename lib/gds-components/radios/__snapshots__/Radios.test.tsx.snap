// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radios calls onChange 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <fieldset
      class="govuk-fieldset"
      id="radio"
    >
      <legend
        class="govuk-fieldset__legend govuk-fieldset__legend--l"
      >
        <h1
          class="govuk-fieldset__heading"
        >
          radio
        </h1>
      </legend>
      <div
        class="govuk-radios"
        data-module="govuk-radios"
      >
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="y"
            name="radios"
            role="radio-input"
            type="radio"
            value="y"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="y"
          >
            Yes
          </label>
        </div>
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="n"
            name="radios"
            role="radio-input"
            type="radio"
            value="n"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="n"
          >
            No
          </label>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`Radios displays 1 radio input for each option 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <fieldset
      class="govuk-fieldset"
      id="radio"
    >
      <legend
        class="govuk-fieldset__legend govuk-fieldset__legend--l"
      >
        <h1
          class="govuk-fieldset__heading"
        >
          radio
        </h1>
      </legend>
      <div
        class="govuk-radios"
        data-module="govuk-radios"
      >
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="y"
            name="radios"
            role="radio-input"
            type="radio"
            value="y"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="y"
          >
            Yes
          </label>
        </div>
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="n"
            name="radios"
            role="radio-input"
            type="radio"
            value="n"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="n"
          >
            No
          </label>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;
